jspsych-uil-randomization.js

```diff
@@ -32,6 +32,113 @@ else
+     *                             item can be appended to randomized. The keys
+     *                             need to be a key in the item parameter and
+     *                             the value is the maximum number allowable
+     *                             subsequent item[key] with the same value.
+     * @param item {object} The object which is intended to be appended. The
+     *                      constraints determine whether it fits.
+     *
+     * @return true if the item may be appended to randomized without violating
+     *         the constraints false otherwise.
+     */
+    function allowPushItem (
+        randomized,
+        constraints,
+        item
+    ) {
+        for (const [key, max] of Object.entries(constraints)) {
+            if (max < 1) {
+                throw new RangeError(
+                    "max is < 1; it's impossible to fit less than one repeating items"
+                );
+            }
+            let value = item[key];
+            if (value === undefined) {
+                throw ReferenceError(
+                    `There is no key "${key}" in item "${item}"`
+                );
+            }
+
+            let selection = randomized.slice(-max);
+            selection.push(item);
+
+            let count = 0;
+            selection.forEach(stimulus => count += stimulus[key] === value);
+            if (count > max)
+                return false;
+        }
+        return true;
+    }
+
+    /**
+     * Randomize the input stimuli, according to the given constraints.
+     *
+     * A new list of stimuli will be returned if reasonably possible.
+     * The constraints are an object with a key that must also be present in
+     * the original stimuli. The value of the belonging to the key, denotes
+     * how many items with item[key] may have the same value in a row.
+     *
+     * @param original_stimuli {array.<object>} The unrandomized stimuli
+     * @param constraints {object} The constraints to determine
+     * how many items with the same value may be appended in a row.
+     * @param nth_try {number} The nth attempt to randomize the stimuli [0, max_tries)
+     * @param max_tries {number} The total number of attempts to randomize the stimuli.
+     *
+     * @returns {null|[]} The randomized order. Null if randomization
+     * failed, in which case an error will have been logged to the console
+     */
+    function randomizePrivate(
+        original_stimuli,
+        constraints,
+        nth_try,
+        max_tries
+    ) {
+        if (max_tries < 1) {
+            throw new RangeError("max_tries is < 1");
+        }
+        if (nth_try < 0) {
+            throw new RangeError("nth_try is < 0");
+        }
+
+        if (nth_try >= max_tries) {
+            console.log(
+                "Unable to randomize according the constraints, "       +
+                "Perhaps it is a good idea to make the constraints "    +
+                "less strict."
+            );
```

Comment:
I'd say 'console.warning' is more apropiate here. .log is more for informational, but this message is definetly more than informational

Reply:
Agreed!

Reply:
I now see, that the original code even used console.error(), that might also be appropriate.

Reply:
Yeah when I was thinking it over just now I thought the same thing, as it's a show-stopping problem

Reply:
While I like your due diligence, you don't _have_ to comment the fixing commits in such tiny code reviews like this. (Unless it helps you of course)

Reply:
(Btw, if you're doing it to mark it as 'done', you can also use the 'Resolve conversation' button on the bottom)